vllm/v1/worker/gpu_model_runner.py

```diff
@@ -103,6 +103,41 @@
 logger = init_logger(__name__)
 
 
+# Wrapper for ModelRunnerOutput to support overlapped execution.
+class AsyncGPUModelRunnerOutput(AsyncModelRunnerOutput):
+
+    def __init__(
+        self,
+        model_runner_output: ModelRunnerOutput,
+        sampled_token_ids: torch.Tensor,
+        invalid_req_indices: list[int],
+        async_output_copy_stream: torch.cuda.Stream,
+    ):
+        self.model_runner_output = model_runner_output
+        self.sampled_token_ids = sampled_token_ids
+        self.invalid_req_indices = invalid_req_indices
+        self.async_output_copy_stream = async_output_copy_stream
+
+    def copy_to_host(self) -> ModelRunnerOutput:
+        """Copy the device tensors to the host and return a ModelRunnerOutput.
+        
+        This function blocks until the copy is finished.
+        """
+        default_stream = torch.cuda.current_stream()
+        with torch.cuda.stream(self.async_output_copy_stream):
+            self.async_output_copy_stream.wait_stream(default_stream)
```

Comment:
It would be better to use an event here I think, or else we could be waiting for unrelated work corresponding to subsequent step
```suggestion
        with torch.cuda.stream(self.async_output_copy_stream):
            self._data_ready.wait()
```

and then at the end of the constructor you can add
```python
        self._data_ready = torch.cuda.default_stream().record_event()
```

Reply:
**OR** we could initiate the copy in the constructor and instead have an event for the copy completion.  This might be nicer on second thoughts:
```python
        # __init__() ...
        self.model_runner_output = model_runner_output
        self.invalid_req_indices = invalid_req_indices
        self._ready_event = torch.cuda.Event(blocking=True)

        default_stream = torch.cuda.current_stream()
        with torch.cuda.stream(async_output_copy_stream):
            async_output_copy_stream.wait_stream(default_stream)
            self.sampled_token_ids_cpu = sampled_token_ids.to(
                'cpu', non_blocking=True)
            self._ready_event.record()
            sampled_token_ids.record_stream()  # not sure if async_output_copy_stream needs to be passed as an arg to this

    def copy_to_host(self) -> ModelRunnerOutput:
        """Copy the device tensors to the host and return a ModelRunnerOutput.
        
        This function blocks until the copy is finished.
        """
        self._ready_event.synchronize()
        valid_sampled_token_ids = self.sampled_token_ids_cpu.tolist()
        for i in self.invalid_req_indices:
            valid_sampled_token_ids[i].clear()
        # ...
```

Reply:
(actually `blocking=True` probably isn't needed here)

Reply:
I've decided to go with the second approach, but without the event: since the constructor is now running in the gpu model runner thread, I think the call to `async_output_copy_stream.wait_stream(default_stream)
` should be sufficient, and it will never capture any work from the subsequent iteration. 

Please let me know if I'm missing an edge case here.

Reply:
I see what you mean, it does seem 'correct' that way. I'll change it.